### PR TITLE
[YOLOv8] Step 0/? Addition of SPP and SPPF Layer

### PR DESCRIPTION
**Pull request by IMvision12**
Added Spatial Pyramid Pooling (SPP) and SPPF layer.

Fixes: #1263
SPP paper : https://arxiv.org/abs/1406.4729
SPPF layer was created by Glenn Jocher which is an optimized version of SPP.

Question: should I create series of many PRs for adding layer 1 by 1 or single PR for everything?

@ianstenbit @LukeWood

### CODE REVIEW

**Comment by IMvision12** [comment]
@ianstenbit Yeah I am working on colab, will share it

File: keras_cv/models/object_detection/yolov8/__internal__/__init__.py [hunk 1]
```diff
@@ -0,0 +1,13 @@
+# Copyright 2023 The KerasCV Authors
```
Comment (LukeWood) [inline keras_cv/models/object_detection/yolov8/__internal__/__init__.py:1]: We can stop using the __internal__ structure --- you can just place this in `keras_cv/models/object_detection/yolov8`

we have a change coming in that will make our API generated instead of dictated by file structure, so this is no longer needed.  Thanks!

File: keras_cv/models/object_detection/yolov8/__internal__/layers/sppf.py [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import tensorflow as tf
+from tensorflow.keras import layers
+
+
+@tf.keras.utils.register_keras_serializable(package="keras_cv")
+class SPP(layers.Layer):
```
Comment (LukeWood) [inline keras_cv/models/object_detection/yolov8/__internal__/layers/sppf.py:20]: Remove unnecessary abbrevaiations:

```
SpatialPyramidPooling
```

File: keras_cv/models/object_detection/yolov8/__internal__/layers/sppf.py [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+    """
+    Performs Spatial Pyramid Pooling.
```
Comment (LukeWood) [inline keras_cv/models/object_detection/yolov8/__internal__/layers/sppf.py:22]: Please provide a description of what this operation means!

File: keras_cv/models/object_detection/yolov8/__internal__/layers/sppf.py [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+    """
+    
+    def __init__(self, filters, kernels=(5, 9, 13), **kwargs):
```
Comment (LukeWood) [inline keras_cv/models/object_detection/yolov8/__internal__/layers/sppf.py:25]: in the docstring provide an args list describing the function of each of these.

File: keras_cv/models/object_detection/yolov8/__internal__/layers/sppf.py [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+            "filters": self.filters,
+            "kernels": self.kernels,
+            "num_filters": self.num_filters,
+        }
+        base_config = super().get_config()
+        return dict(list(base_config.items()) + list(config.items()))
+
+
+@tf.keras.utils.register_keras_serializable(package="keras_cv")
+class SPPF(layers.Layer):
```
Comment (LukeWood) [inline keras_cv/models/object_detection/yolov8/__internal__/layers/sppf.py:58]: Same here - "FastSpatialPyramidPooling"

Comment (LukeWood) [reply keras_cv/models/object_detection/yolov8/__internal__/layers/sppf.py:58]: same comment w/ constructor arguments

File: keras_cv/models/object_detection/yolov8/__internal__/layers/sppf.py [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+        self.filters = filters
+        self.kernel = kernel
+        self.num_filters = filters // 2
+        self.conv1 = layers.Conv2D(
+            filters=self.num_filters, kernel_size=1, strides=1
+        )
+        self.conv2 = layers.Conv2D(
+            filters=self.filters, kernel_size=1, strides=1
+        )
+        self.module = layers.MaxPool2D(
```
Comment (LukeWood) [inline keras_cv/models/object_detection/yolov8/__internal__/layers/sppf.py:74]: lets rename module here because now it is only one layer

File: keras_cv/models/object_detection/yolov8/__internal__/layers/sppf.py [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+    """
+    Performs Spatial Pyramid Pooling.
+    """
+    # Spatial Pyramid Pooling - Fast (SPPF) layer for YOLOv5 by Glenn Jocher
+    def __init__(self, filters, kernel=5, **kwargs):
+        super().__init__(**kwargs)
+        self.filters = filters
+        self.kernel = kernel
+        self.num_filters = filters // 2
+        self.conv1 = layers.Conv2D(
```
Comment (LukeWood) [inline keras_cv/models/object_detection/yolov8/__internal__/layers/sppf.py:68]: How about `input_conv` and `output_conv` as names?

File: keras_cv/models/object_detection/yolov8/__internal__/layers/sppf.py [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+        return dict(list(base_config.items()) + list(config.items()))
+
+
+@tf.keras.utils.register_keras_serializable(package="keras_cv")
+class SPPF(layers.Layer):
+    """
+    Performs Spatial Pyramid Pooling.
+    """
+    # Spatial Pyramid Pooling - Fast (SPPF) layer for YOLOv5 by Glenn Jocher
+    def __init__(self, filters, kernel=5, **kwargs):
```
Comment (LukeWood) [inline keras_cv/models/object_detection/yolov8/__internal__/layers/sppf.py:63]: lets rename `kernel` to `kernel_size`

File: keras_cv/models/object_detection/yolov8/__internal__/layers/sppf.py (hunk 4 above)
Comment (LukeWood) [inline keras_cv/models/object_detection/yolov8/__internal__/layers/sppf.py:25]: kernels => kernel_sizes perhaps?

**Review by LukeWood — CHANGES_REQUESTED** [review]
Thanks for the PR!  I left a first batch of changes on the PR.

**Comment by IMvision12** [comment]
Hi @ianstenbit , I have converted the yolov8 architecture, I need to verify if it's correct or not, so how can i do that..

**Comment by IMvision12** [comment]
@LukeWood @ianstenbit are we going to add a training script too? till now I have completed with model and loss currently working on training!!!

**Comment by ianstenbit** [comment]
> @LukeWood @ianstenbit are we going to add a training script too? till now I have completed with model and loss currently working on training!!!

Hi @IMvision12 -- this is exciting! 
A training script would be great, yes. Ideally this would start as a Colab that we can later add as an example in KerasCV.

In order to verify correctness, we should find a reference implementation and see if we can port its weights and verify numerical equivalence.

**Comment by haifeng-jin** [comment]
Hi @IMvision12 , thank you so much for the contribution!
We are prioritizing the YOLO model in KerasCV.
I will be assisting you in merging your PR.

Would you please update us with the current status of this PR?

**Comment by IMvision12** [comment]
Hii, @haifeng-jin, This PR is probably ready to merge, but as @ianstenbit proposed, we need to first build a complete functional model before we can start merging PRs one by one, so I'm working on creating a full implementation of yolov8 here on my repo: https://github.com/IMvision12/YOLOv8-tf, currently I am working on loss, modeling part is completed.

**Comment by haifeng-jin** [comment]
Thanks @IMvision12 ! Looking forward to it.

**Comment by haifeng-jin** [comment]
Hi @IMvision12, would you please let us know a estimated timeline?
We would like to coordinate this line of work with our overall release schedule of KerasCV.
Thank you!

**Comment by IMvision12** [comment]
Hello, @haifeng-Jin, YOLOV8, and other OD and segmentation models align with my GSoC timeline, so if possible can you please go through the timeline in the proposal?

**Comment by jbischof** [comment]
@IMvision12 thanks so much for your contribution! We haven't accepted any GSoC proposals yet and will only have one slot for KerasCV. Would you still be interested in contributing this model outside the GSoC program?

**Comment by IMvision12** [comment]
Yes @jbischof

**Comment by IMvision12** [comment]
#1675
